### ISSUE TITLE
`indent` not playing nice with comma-first style

### ISSUE THREAD

**Issue by zyklus** [eslint/eslint#3456]
```js
{
    foo : 123
  , bar : 456
}
```

throws an error with `indent` expecting 2 spaces instead of 4.  However I prefer to align my keys like this instead of either of these:

```js
{
  foo : 123
  , bar : 456
}
```
```js
{
  foo   : 123
  , bar : 456
}
```

Is there any way to get indent to be quiet here but keep it on everywhere else?  This didn't happen in v0.24, currently on v1.2.0

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by BYK** [comment]
@zyklus - can you share your configuration for this rule too? I suspect your indent level is `2` because when I test this with `4` it works.

**Comment by BYK** [comment]
And by "it works" I mean it doesn't give me any warnings.

**Comment by zyklus** [comment]
@BYK -- Sorry, yes it's set to `[2, 2]`

**Comment by BYK** [comment]
@zyklus - then I think it makes sense for the rule to error. I don't think this is something that should be fixed.

**Comment by zyklus** [comment]
@BYK -- It's a conflict with a comma-first style and, similar to options like `const`, allows keys to be properly aligned:

```js
const a = 1,
      b = 2,
      c = 3;
```

If eslint supports a comma-first style, I believe it should be fixed :)

**Comment by BYK** [comment]
@zyklus - it doesn't conflict with comma first style actually because it works when you set the indent to `4` with that same code. The issue here is that code not adhering to 2-space indent rules really. If it gave a warning with the following example, I may have considered this as a bug:

```js
var a = {
  some: 1
, name: 2
};

var b = {
  same: 3
 ,like: 4
};

**Comment by zyklus** [comment]
I don't consider this "proper" indentation like you're suggesting though:

```js
config = extend({
  params       : args.shift()
, description  : args.shift()
, handler      : args.shift()
, defaultValue : args.shift()
}, config || {});
```

**Comment by zyklus** [comment]
To clarify -- I acknowledge that the code doesn't adhere to a 2-spacing rule, but at the same time I believe it's "properly" spaced given a comma-first style.

So whether my original issue of "indent is broken" is correct, I believe there should be _some_ way to make the code pass linting while simultaneously keeping the 2-space rule for all other code.

**Comment by BYK** [comment]
>I believe there should be some way to make the code pass linting while simultaneously keeping the 2-space rule for all other code.

Well, you can disable the rule for that part :dancer: 

Anyways, I see what you say but still disagree with you. I also think trying to support this would make the already complicated rule even more complicated. That said I'm just one person so let's see what others think. @eslint/eslint-team, please weigh in :)

**Comment by zyklus** [comment]
> Well, you can disable the rule for that part :dancer:

how?  I don't see a way to disable it specifically for objects.

**Comment by BYK** [comment]
>Or are you suggesting wrapping every object block in comments that disable the rule? :rage:

I was suggesting that (and sorry for the dancer emoji, I meant to put a `:D` there).

**Comment by gyandeeps** [comment]
I agree with @BYK suggestion of diabling the rule here.
@zyklus Here is the link which explains how to disable rules: http://eslint.org/docs/user-guide/configuring#configuring-rules

**Comment by ilyavolodin** [comment]
@zyklus While we would love to support this style, it's just not feasible for us to support every indentation style under the sun. Indent is already the most complicated rule that we have, and adding every additional style makes it even more complex. I would also suggest disabling this rule for declarations.

**Comment by kentor** [comment]
Is there a way to specify in the indent rule to just turn off checking variable declarations? This has been preventing me from upgrading from 1.0.0, unless I turn off checking indentations entirely. I don't want to add eslint-ignore directives to a lot of existing code.

**Comment by zyklus** [comment]
I'm with @kentor on this.  Though I don't think my style is that uncommon (there was a chart floating around a while ago that said comma-first was around 20%), I concede that supporting every style is unfeasible.  It would be nice to have a way to disable this rule contextually.  In my case for object literals.  Though yes, it's _possible_ to disable this by surround blocks with comments, that's hideous.  If it was a one-off, fine, but adding comments to disable this rule in 100 different places isn't an acceptable "solution".

**Comment by BYK** [comment]
Adding a new option "exceptions" that takes a node list or something to turn checking off for certain nodes looks doable but I still think the style you are using is inconsistent indentation and trying to make the linter accept that (even through skipping object literals everywhere) is not a good practice.

Keep in mind that this would cover all `ObjectExpression` nodes which may be nested so don't need the exception at all. Looks like this is only a problem when you have all of the following conditions:

- comma-first style
- a top level var-statement
- indent level: 2 spaces

**Comment by nzakas** [comment]
I like the `exceptions` idea. That will buy us some runway for styles we can't figure out how to support.

**Comment by ilyavolodin** [comment]
I think we should think about making reusable "exception by node type" functionality. Indent already has a use-case of this with additional indentation.

**Comment by zyklus** [comment]
> I still think the style you are using is inconsistent indentation

@BYK - Of course it is, but it's a matter of priorities.  It is _impossible_ to have both consistent indentation and aligned keys in this case with a comma-first style.  As the entire point here is about style, to me aligned keys come first in priority.

+1 to general exceptions, or possibly a way to re-configure rules by node type, though I've no idea how much that would complicate either the code base or config files.

**Comment by keis** [comment]
The use of semicolons also seems to play a part

this fails on line 4
```
var foo = 1
    ,bar = 2

var baz = 3
```

while this passes 
```
var foo = 1
    ,bar = 2;

var baz = 3
```

using this configuration
```
{
  "rules": {
    "comma-style": [2, "first"],
    "indent": 2
  }
}
```

**Comment by BYK** [comment]
@keis - This is a different issue addressed in #3515
